pyproject.toml

```diff
@@ -0,0 +1,52 @@
+[project]
+name = "OpenImageIO"
+version = "2.4.0.dev1"
+description = "Reading, writing, and processing images in a wide variety of file formats, using a format-agnostic API, aimed at VFX applications."
+authors = [
+    {name = "Larry Gritz", email = "lg@larrygritz.com"},
+]
+dependencies = [
+    "setuptools>=68.2.2",
+]
+requires-python = ">2.7, !=3.0.*, !=3.1.*, !=3.2.*, !=3.3.*, !=3.4.*, !=3.5.*, !=3.6.*, !=3.9.0"
```

Comment:
Is there a particular reason to exclude 3.9.0 exactly?  Also, would attempting to use python 2.7.18 pass this check? If so, I have a meta question/comment around if we want to continue to support and encourage such nonsense :)

Reply:
3.9 is known to be buggy in when used with Pybind11.

Reply:
And no, I don't think we should support 2.7. We'll need at least Pybind11 2.10, which doesn't support python 2.7.

Reply:
VFX Platform 2022 uses Python 3.9, so it's definitely in the mix of things we support, and I don't recall any pybind11 problems (for OIIO anyway).

Reply:
Removed all Python version restrictions. Definitely not going to support 2.7 or anything older than 3.9 at this point, but I'm not sure yet if we need to be explicit about that here.

Reply:
2.7 is definitely fine to remove, we'll surely remove all 2.7 support by the time OIIO 2.6 is fully baked next summer.

"Nothing older than 3.9" means "people can't pip install a version that will work with any VFX platform year prior to 2022. Which may be fine? But it's a choice we should make with eyes open.

Reply:
I would still support 3.7 and 3.8. 3.7 is still widely used by the industry and all ASWF projects ship wheels for 3.7, 3.8, 3.9, 3.10 and 3.11. By experience, supporting 3.7 doesn't add any extras work or burden when using Pybind11.

Reply:
Also 3.9.0 was released Oct. 5, 2020. 3.9.18 is the [latest point release](https://www.python.org/downloads/), released Aug 24, 2023. So it's probably not practical to exclude a single point release when there are 17 releases between it and the latest point release. If anything, we'd probably exclude `<3.9.18,<3.10.13,<3.11.6,<3.12.0` and so on.

Reply:
We don't want to cut off users who need particular older versions, but at the same time, if the only people who care about those versions are big studios or products that build from source anyway, then maybe those are indeed cases we should exclude by design.

I think we need to start with clear first principles statement here: which group of users are we aiming to cater to with a pip install solution, and what versions of python does that pool of users care about?

Reply:
Usually, the target group for wheels is anyone not part of a big studio, students, scientists, etc. All people that don't necessarily either have control over the python versions they have, or don't necessarily know that python 3.7 is EOL. Usually, when users pip install, they expect things to just work. By experience, if we don't support 3.7 or 3.11, we'll get support requests all the time. Users will ask for them. 3.7 because DCCs use it, and 3.11 because users want it. Requests for 3.12 will also start to arrive soon (we already had one in OTIO).

Reply:
This is the roadmap, right? https://vfxplatform.com/

In that case, you could go all the way back to VFX 2017 and add support for whatever Python was in use then, but for the purposes of the sprint I would say since CY2024 remains on VFX 2022 that we support 3.8 and higher in the next OIIO release because that's what's supported by Python at the time of the release.

The further back you go with Python versions, the more work you may end up doing. But you are correct you need to know what version of Python is in use and you may want to include that information on the VFX roadmap. I can picture a "worst case" scenario where everyone uses Python 3.7 with VFX 2022. I wouldn't like it, but we could probably focus on that 3.7 only in that case :smile: .

Lastly, this may be slightly premature as it will be nice to get just some useful wheels built in any version over the next 48 hours or so!

Reply:
This may help: https://docs.google.com/spreadsheets/d/1EwRlz5ZYObEOdBfIk8iTX5thlpTyEAfp3bxOgAfFOiU/edit#gid=1592527777

That shows (near the bottom) what VFX Platform years the different studios are using. Prior to 2020 is definitely not needed, and I think even 2020 (maybe 2021) is on the way to being phased out. This also matches with VFX Platform's own recommendation of trying to support up to 3 years back from the present.

Reply:
I'll put it in even more stark terms: my own studio will definitely be transitioned away from Python 2.7 by the end of this year, I hope. (And onto an assortment of 3.7, 3.9, 3.10, depending on the specific one of the several DCCs we are depending on.)

Reply:
In terms of priority, it's likely that the users who need a simple "pip install" approach most desperately (i.e., want to avoid building anything from source) are very likely on a very recent python. So that's probably top priority. But I'm playing devil's advocate here to make sure we don't back ourselves into a corner, design wise, that will make it difficult to eventually support python versions as far back as the VFX years we claim to support.

Reply:
I do agree that we should definitely not support python older than 3.7. Whether 3.7 or 3.9 is the minimum we want to support for "pip install" is reasonable for debate, and I don't have a firm opinion on it, other than that we should choose among them for a principled reason.

Reply:
The code base already supports 3.7 and up thanks to Pybind11, so building for 3.7 isn't more difficult than building for 3.11. 3.12 is a different story though because it's still pretty recent (it's only one week old). In other words, once you get a build running and working wheels for one version, you get the other versions for free without any additional work.